src/dbnode/client/session.go

```diff
@@ -1203,43 +1261,124 @@ func (s *session) fetchTaggedAttemptWithRLock(
 	var (
 		topoMap    = s.state.topoMap
-		op         = s.pools.fetchTaggedOp.Get()
 		fetchState = s.pools.fetchState.Get()
 	)
+	fetchState.nsID = ns // transfer ownership to `fetchState`
+	fetchState.incRef()  // indicate current go-routine has a reference to the fetchState
+
+	// wire up the operation based on the opts specified
+	var (
+		op     op
+		closer func()
+	)
+	switch opts.stateType {
+	case fetchTaggedFetchState:
+		fetchOp := s.pools.fetchTaggedOp.Get()
+		fetchOp.incRef() // indicate current go-routine has a reference to the op
+		closer = func() {
+			fetchOp.decRef() // release the ref for the current go-routine
+		}
+		fetchOp.update(opts.fetchTaggedRequest, fetchState.completionFn)
+		fetchState.ResetFetchTagged(opts.startInclusive, opts.endExclusive,
+			fetchOp, topoMap, s.state.majority, s.state.readLevel)
+		op = fetchOp
+
+	case aggregateFetchState:
+		aggOp := s.pools.aggregateOp.Get()
+		aggOp.incRef() // indicate current go-routine has a reference to the op
+		closer = func() {
+			aggOp.decRef() // release the ref for the current go-routine
+		}
+		aggOp.update(opts.aggregateRequest, fetchState.completionFn)
+		fetchState.ResetAggregate(opts.startInclusive, opts.endExclusive,
+			aggOp, topoMap, s.state.majority, s.state.readLevel)
+		op = aggOp
 
-	fetchState.nsID = nsClone // transfer ownership to `fetchState`
-	fetchState.incRef()       // indicate current go-routine has a reference to the fetchState
-	op.incRef()               // indicate current go-routine has a reference to the op
-	op.update(req, fetchState.completionFn)
+	default:
+		fetchState.decRef() // release fetchState
+		instrument.EmitInvariantViolation(s.opts.InstrumentOptions())
+		return nil, xerrors.NewNonRetryableError(instrument.InvariantErrorf(
+			"unknown fetchState type: %v", opts.stateType))
+	}
 
-	fetchState.Reset(opts.StartInclusive, opts.EndExclusive, op, topoMap, s.state.majority, s.state.readLevel)
 	fetchState.Lock()
 	for _, hq := range s.state.queues {
 		// inc to indicate the hostQueue has a reference to `op` which has a ref to the fetchState
 		fetchState.incRef()
 		if err := hq.Enqueue(op); err != nil {
 			fetchState.Unlock()
-			op.decRef()         // release the ref for the current go-routine
+			closer()
 			fetchState.decRef() // release the ref for the hostQueue
 			fetchState.decRef() // release the ref for the current go-routine
 
 			// NB: if this happens we have a bug, once we are in the read
 			// lock the current queues should never be closed
-			wrappedErr := fmt.Errorf("[invariant violated] failed to enqueue fetchTagged: %v", err)
-			s.log.Errorf(wrappedErr.Error())
+			wrappedErr := xerrors.NewNonRetryableError(fmt.Errorf("failed to enqueue in fetchState: %v", err))
+			instrument.EmitAndLogInvariantViolation(s.opts.InstrumentOptions(), func(l xlog.Logger) {
+				l.Errorf("%v", wrappedErr)
+			})
 			return nil, wrappedErr
 		}
 	}
-
-	op.decRef() // release the ref for the current go-routine
+	closer()
```

Comment:
nit: re-add the `// release the ref for the current go-routine` comment here and above

Reply:
Yeah agreed.